File: examples/cxx/autoDestination/Master.cxx
#include "Master.hpp"
#include "BasicTask.hpp"
#include "FinalizeTask.hpp"

#include <Host.hpp>
#include <HostNotFoundException.hpp>

#include <iostream>
using namespace std;


#ifndef BUFFMAX
#define BUFFMAX 260
#endif

MSG_IMPLEMENT_DYNAMIC(Master, Process);

int Master::main(int argc, char** argv)
{
	char buff[BUFFMAX + 1] = {0};
	int taskCount;		
	double taskComputeSize;		
	double taskCommunicateSize;
	
	cout << "Hello I'm " << getName() << " on " << getHost().getName() << "!" << endl;
		
	cout <<"[" << getName() << ":" << getHost().getName() << "] " << "argc=" << argc << endl;
		
	for (int i = 0; i< argc; i++)	    
		cout << "argv:" << argv[i] << endl;
	
	sscanf(argv[0],"%d", &taskCount);
	sscanf(argv[1],"%lg", &taskComputeSize);
	sscanf(argv[2],"%lg", &taskCommunicateSize);
	
	BasicTaskPtr* basicTasks = new BasicTaskPtr[taskCount];
		
	for (int i = 0; i < taskCount; i++) 
	{
		sprintf(buff,"Task_%d",i);
		basicTasks[i] = new BasicTask(buff, taskComputeSize, taskCommunicateSize);
		memset(buff, 0 , BUFFMAX + 1); 
	}		
	
	int aliasCount = argc - 3;
		
	char** aliases = (char**) calloc(aliasCount, sizeof(char*));
		
	for(int i = 3; i < argc ; i++) 
		aliases[i - 3] = _strdup(argv[i]);
		
	cout <<"[" << getName() << ":" << getHost().getName() << "] " << "Got "<<  aliasCount << " alias(es) :" << endl;
		
	for (int i = 0; i < aliasCount; i++)
		cout <<"[" << getName() << ":" << getHost().getName() << "] " << "\t" << aliases[i] << endl;
		
	cout <<"[" << getName() << ":" << getHost().getName() << "] " << "Got "<< taskCount << " task to process." << endl;
		
	for (int i = 0; i < taskCount; i++) 
	{	
		cout <<"[" << getName() << ":" << getHost().getName() << "] " << "Sending \"" << basicTasks[i]->getName() << "\" to \"" << aliases[i % aliasCount] << "\"" << endl;
		
		/*if((Host::currentHost().getName()).equals((aliases[i % aliasCount].split(":"))[0]))
			cout <<"[" << getName() << ":" << getHost().getName() << "] " << "Hey ! It's me ! ";
		*/
			
		basicTasks[i]->send(aliases[i % aliasCount]);
	}
		
	cout <<"[" << getName() << ":" << getHost().getName() << "] " << "Send completed" << endl;
	
	cout <<"[" << getName() << ":" << getHost().getName() << "] " << "All tasks have been dispatched. Let's tell everybody the computation is over." << endl;
	
	FinalizeTask* finalizeTask;
	
	for (int i = 0; i < aliasCount; i++) 
	{
		finalizeTask = new FinalizeTask();
		finalizeTask->send(aliases[i]);
		
	}
	
	cout <<"[" << getName() << ":" << getHost().getName() << "] " << "Goodbye now!" << endl;

	delete[] basicTasks;
	delete[] aliases;

	delete this;

	return 0;
}

File: examples/deprecated/java/app/masterworker/Master.java
/* Master of a basic master/worker example in Java */

/* Copyright (c) 2006-2020. The SimGrid Team.
 * All rights reserved.                                                     */

/* This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. */

package app.masterworker;
import org.simgrid.msg.Host;
import org.simgrid.msg.Msg;
import org.simgrid.msg.MsgException;
import org.simgrid.msg.Task;
import org.simgrid.msg.Process;

public class Master extends Process {
  public Master(Host host, String name, String[]args) {
    super(host,name,args);
  } 
  public void main(String[] args) throws MsgException {
    if (args.length < 4) {
      Msg.info("Master needs 4 arguments");
      System.exit(1);
    }

    int tasksCount = Integer.parseInt(args[0]);
    double taskComputeSize = Double.parseDouble(args[1]);
    double taskCommunicateSize = Double.parseDouble(args[2]);

    int workersCount = Integer.parseInt(args[3]);

    Msg.info("Hello! My PID is "+getPID()+". Got "+  workersCount + " workers and "+tasksCount+" tasks to process");

    for (int i = 0; i < tasksCount; i++) {
      Task task = new Task("Task_" + i, taskComputeSize, taskCommunicateSize); 
      Msg.debug("Sending \"" + task.getName()+ "\" to \"worker_" + i % workersCount + "\"");
      task.send("worker_"+(i%workersCount));
    }

    Msg.info("All tasks have been dispatched. Let's tell everybody the computation is over.");

    for (int i = 0; i < workersCount; i++) {
      Task task = new Task("finalize", 0, 0);
      task.send("worker_"+(i%workersCount));
    }

    Msg.info("Goodbye now!");
  }
}
